ARM64-SVE: refactor lsra buildHWIntrinsic

a74nh opened the pull request:
Fixes #104842

The logic for hwintrisics has become convoluted. Refactor it, for both SVE and AdvSimd.

Add functions to get the operand (if any) for each requirement - delay slot, consecutive registers, address, etc.

Then use a simple for loop to iterate through each operand and build depending on which requirements match for that operand.

Tested by using stress_test.py on the entire HardwareIntrinsics_Arm set.

a74nh committed:
Fixes from other PRs to be removed

a74nh committed:
Add BuildConditionalSelectHWIntrinsic()

a74nh committed:
Add GetRMWOp()

a74nh committed:
Use GetDelayFreeOp() in BuildConditionalSelectWithEmbeddedOp()

a74nh committed:
simplify op2 handling

a74nh committed:
Add getVectorAddrOperand()

a74nh committed:
Add getConsecutiveRegistersOperand

a74nh committed:
Add BuildOperand()

a74nh committed:
Use BuildOperand for op1

a74nh committed:
Add buildHWIntrinsicImmediate

a74nh committed:
Add getOperandCandidates()

a74nh committed:
Remove BuildOperand()

a74nh committed:
remove delayFreeMultiple

a74nh committed:
Fix formatting

a74nh commented:
@kunalspathak - Still WIP. For now ignore anything outside of `lsraarm64.cpp` and `lsra.hpp`. All the other changes are from other PRs and will be removed was those have been merged.

a74nh committed:
Replace BuildConditionalSelectWithEmbeddedOp() with BuildEmbeddedOperandUses()

a74nh committed:
Use BuildHWIntrinsicImmediate for conditional select

a74nh committed:
Remove failing unary tests

a74nh committed:
Remove IsRMW

a74nh committed:
Revert "Fixes from other PRs to be removed"

a74nh committed:
Move functions

a74nh committed:
Move functions

a74nh committed:
Fix opNum type

a74nh commented:
This PR is ready now.

Requires #107084, #107180 and a workaround for #107537 in order for all the hwintrinsic tests to pass.

Apologies, this is a large change to review, and the github diff is confused about functions I haven't touched. Probably best starting a review from the new version of `BuildHWIntrinsic()` 

I recommend this is not merged until after we've gone past the Net9 RC2 deadline.

@dotnet/arm64-contrib

I'll do a spmidiff next.

a74nh committed:
Revert "Remove failing unary tests"

kunalspathak commented:
I expected this to be no diff changes, but looks like [it is not](https://dev.azure.com/dnceng-public/public/_build/results?buildId=802210&view=ms.vss-build-web.run-extensions-tab). Can you please double check the source of differences?

a74nh commented:
> I expected this to be no diff changes, but looks like [it is not](https://dev.azure.com/dnceng-public/public/_build/results?buildId=802210&view=ms.vss-build-web.run-extensions-tab). Can you please double check the source of differences?

This looks like it's all to `LoadAndInsertScalar`.

tldr: there is a bug in HEAD where `getVectorAddrOperand()` is not used for `op3`

Long version:

There are multiple versions of `LoadAndInsertScalar` because it needs to handle variants with multiple op1/return values
All of them have an address operand in `op3`

```
public static unsafe (Vector128<byte> Value1, Vector128<byte> Value2) LoadAndInsertScalar((Vector128<byte>, Vector128<byte>) values, [ConstantExpected(Max = (byte)(15))] byte index, byte* address);

public static unsafe Vector128<uint> LoadAndInsertScalar(Vector128<uint> value, [ConstantExpected(Max = (byte)(3))] byte index, uint* address);
```

In HEAD, for the multiple register versions, `NI_AdvSimd_LoadAndInsertScalarVectorXxX`, it has special handling:

```
else if (HWIntrinsicInfo::NeedsConsecutiveRegisters(intrin.id))
....
            case NI_AdvSimd_LoadAndInsertScalarVector64x2:
            case NI_AdvSimd_LoadAndInsertScalarVector64x3:
            case NI_AdvSimd_LoadAndInsertScalarVector64x4:
            case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x2:
            case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x3:
            case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x4:
            {
                assert(intrin.op2 != nullptr);
                assert(intrin.op3 != nullptr);
                assert(isRMW);
                if (!intrin.op2->isContainedIntOrIImmed())
                {
                    srcCount += BuildOperandUses(intrin.op2);
                }

                assert(intrinsicTree->OperIsMemoryLoadOrStore());
                srcCount += BuildAddrUses(intrin.op3);
                buildInternalRegisterUses();
                FALLTHROUGH;
            }
```
Note that `BuildAddrUses()` is used for `op3`

For the single register variant, `NI_AdvSimd_LoadAndInsertScalar`, it doesn't have `NeedsConsecutiveRegisters` so falls into the generic `op2` handling code, before falling into the generic `op3` handling code, which does:

```
        if (intrin.op3 != nullptr)
        {
            SingleTypeRegSet candidates = lowVectorOperandNum == 3 ? lowVectorCandidates : RBM_NONE;

            if (isRMW)
            {
                srcCount += BuildDelayFreeUses(intrin.op3, (tgtPrefOp2 ? intrin.op2 : intrin.op1), candidates);
            }
            else
            {
                srcCount += BuildOperandUses(intrin.op3, candidates);
            }
```
This is wrong - it should be using `BuildAddrUses()` for `op3`.


In my PR, `getVectorAddrOperand()` will correctly return `op3` for all `LoadAndInsertScalarVector` and then the main `for` loop in `BuildHWIntrinsic()` will correctly call `BuildAddrUses()`

a74nh committed:
Remove cases from getDelayFreeOperand that are handled by default

a74nh committed:
Merge main

a74nh commented:
> tldr: there is a bug in HEAD where `getVectorAddrOperand()` is not used for `op3`

In addition, `op1` will be called with `BuildAddrUses()`, which is also wrong.


That happens in:
```
else if (intrinsicTree->OperIsMemoryLoadOrStore())
        {
            srcCount += BuildAddrUses(intrin.op1);
        }
```

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1583,448 +1719,66 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+    BuildHWIntrinsicImmediate(embeddedOpNode, intrinEmbedded);
 
-    if ((intrin.category == HW_Category_SIMDByIndexedElement) && (genTypeSize(intrin.baseType) == 2))
+    if (embeddedDelayFreeOp != nullptr)
```
```suggestion
    if (embeddedDelayFreeOp == nullptr) {  srcCount += BuildOperandUses(embeddedOpNode); return srcCount; }
```

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1583,448 +1719,66 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+            GenTree* node = embeddedOpNode->Op(argNum);
 
-            if (node == prefUseNode)
+            if (node == embeddedDelayFreeOp)
```
can you add a `TODO-CQ` in `else` part that this should be revisited and improved?

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1385,9 +1385,104 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+    GenTree* delayFreeOp = getDelayFreeOperand(intrinsicTree);
+
+    // Determine whether this is an operation where one of the ops is an address
+    GenTree* addrOp = LinearScan::getVectorAddrOperand(intrinsicTree);
```
```suggestion
    GenTree* addrOp = getVectorAddrOperand(intrinsicTree);
```

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1385,9 +1385,104 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+            assert(candidates == RBM_NONE);
+
+            // Some operands have consective op which is also a delay free op
+            srcCount += BuildConsecutiveRegistersForUse(operand, delayFreeOp);
```
We also seems to call `buildInternalRegisterUses()` for consecutive registers. Are we missing it here?

kunalspathak replied:
Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.

- BuildConsecutiveRegistersForUse
- buildInternalRegisterUses
-  BuildDef
- buildInternalRegisterUses
-  BuildDef

a74nh replied:
> We also seems to call `buildInternalRegisterUses()` for consecutive registers. Are we missing it here?

All intrinsics will call `buildInternalRegisterUses()` after the `for` loop, before building the destination.

a74nh replied:
> Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.
> 
> * BuildConsecutiveRegistersForUse
> * buildInternalRegisterUses
> * BuildDef
> * buildInternalRegisterUses
> * BuildDef

In old code:
* op1 - BuildUse (because tgtPrefUse is set, which is because isRMW)
* op2 - BuildConsecutiveRegistersForUse
* op3 - BuildDelayFreeUses
* buildInternalRegisterUses
* BuildDef

In new code:

- delay free = op1 (because isRMW)
- addr = nullptr
- consecutive = op2
- dest consecutive = false
- embedded = nullptr

* BuildHWIntrinsicImmediate (which is a nop)
* op1 - BuildUse (because delayFreeOp == op1)
* op2 - BuildConsecutiveRegistersForUse (because consecutive == op2)
* op3 - BuildDelayFreeUses (because delay free != nullptr)
* buildInternalRegisterUses
* BuildDef

a74nh replied:
You're missing the thing I always miss....

> Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.
> 
> * BuildConsecutiveRegistersForUse
> * buildInternalRegisterUses
> * BuildDef

there is a `return srcCount;` here (line 1934)

Which means it never does:

> * buildInternalRegisterUses
> * BuildDef

kunalspathak reviewed (changes requested):
Overall looks much cleaner. We should also run `jitstress` and other outerloop legs before merging. I can do it once we are done with addressing the feedback.

a74nh committed:
review cleanups

a74nh commented:
Got some asmdiffs for the SVE tests. Spotted two differences, and one of them is due to issues in HEAD.

I'll raise PRs to fix these (plus one for LoadAndInsertScalar), and then rebase this once merged. I'd like there to be no asmdiff differences in this PR


```
./4546.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_BitwiseClear_long RunClassFldScenario() this (FullOpts)
./4000.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_AddSaturate_byte RunBasicScenario_Load() this (FullOpts)
./4130.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_sbyte RunClassFldScenario() this (FullOpts)
./27034.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_byte RunBasicScenario_Load() this (FullOpts)
./22619.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_sbyte RunClassFldScenario() this (FullOpts)
./22615.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_sbyte RunBasicScenario_Load() this (FullOpts)
./4170.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_int RunBasicScenario_Load() this (FullOpts)
./26818.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_int RunClassFldScenario() this (FullOpts)
./26730.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_uint RunClassFldScenario() this (FullOpts)
./4026.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_AddSaturate_ushort RunClassFldScenario() this (FullOpts)
./22659.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_int RunBasicScenario_Load() this (FullOpts)
./4280.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_ulong RunBasicScenario_Load() this (FullOpts)
./4258.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_uint RunBasicScenario_Load() this (FullOpts)
./4192.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_long RunBasicScenario_Load() this (FullOpts)
./26726.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_uint RunBasicScenario_Load() this (FullOpts)
./26906.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_uint RunClassFldScenario() this (FullOpts)
./3481.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_long RunBasicScenario_Load() this (FullOpts)
./26642.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_int RunClassFldScenario() this (FullOpts)
./3547.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_uint RunBasicScenario_Load() this (FullOpts)
./4608.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_BitwiseClear_uint RunBasicScenario_Load() this (FullOpts)
./26946.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_sbyte RunBasicScenario_Load() this (FullOpts)
./27082.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_uint RunClassFldScenario() this (FullOpts)
./3529.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_ushort RunClassFldScenario() this (FullOpts)
./26968.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_short RunBasicScenario_Load() this (FullOpts)
./27012.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_long RunBasicScenario_Load() this (FullOpts)
./22681.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_long RunBasicScenario_Load() this (FullOpts)
./22769.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_ulong RunBasicScenario_Load() this (FullOpts)
./4218.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_byte RunClassFldScenario() this (FullOpts)
./27056.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_ushort RunBasicScenario_Load() this (FullOpts)
./4520.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_BitwiseClear_int RunBasicScenario_Load() this (FullOpts)
./26704.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_ushort RunBasicScenario_Load() this (FullOpts)
./26994.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_int RunClassFldScenario() this (FullOpts)
./4214.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_And_byte RunBasicScenario_Load() this (FullOpts)
./22747.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_uint RunBasicScenario_Load() this (FullOpts)
./22729.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_ushort RunClassFldScenario() this (FullOpts)
./26792.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_short RunBasicScenario_Load() this (FullOpts)
./3415.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_sbyte RunBasicScenario_Load() this (FullOpts)
./3441.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_short RunClassFldScenario() this (FullOpts)
./26554.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_float RunClassFldScenario() this (FullOpts)
./26902.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_uint RunBasicScenario_Load() this (FullOpts)
./4634.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_BitwiseClear_ulong RunClassFldScenario() this (FullOpts)
./22637.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_short RunBasicScenario_Load() this (FullOpts)
./3507.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_byte RunClassFldScenario() this (FullOpts)
./26616.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Subtract_short RunBasicScenario_Load() this (FullOpts)
./26880.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_ushort RunBasicScenario_Load() this (FullOpts)
./22707.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_byte RunClassFldScenario() this (FullOpts)
./22703.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_byte RunBasicScenario_Load() this (FullOpts)
./3459.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_int RunBasicScenario_Load() this (FullOpts)
./22641.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Or_short RunClassFldScenario() this (FullOpts)
./3503.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_byte RunBasicScenario_Load() this (FullOpts)
./26990.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_int RunBasicScenario_Load() this (FullOpts)
./3419.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Add_sbyte RunClassFldScenario() this (FullOpts)
./26814.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_SubtractSaturate_int RunBasicScenario_Load() this (FullOpts)
./27078.dasm:1 :   ; Assembly listing for method JIT.HardwareIntrinsics.Arm._Sve.SimpleBinaryOpTest__Sve_Xor_uint RunBasicScenario_Load() this (FullOpts)
```

a74nh committed:
Simplify masks in getOperandCandidates()

a74nh committed:
Remove IsMaskedOperation()

a74nh committed:
Check for optional embedded masks in getDelayFreeOperand

a74nh commented:
Latest version fixes up the diffs there were caused by this PR.
Once #107786 and #107791 are merged there should be no remaining diffs in this PR.

kunalspathak commented:
> Latest version fixes up the diffs there were caused by this PR. Once #107786 and #107791 are merged there should be no remaining diffs in this PR.

Let's rebase this PR once the above mentioned PRs are merged to confirm there is zero asmdiff.

a74nh committed:
Merge main

Change-Id: Id60f884b7281a9fae85a948a361511656c91357e

a74nh committed:
Merge main

a74nh committed:
Only call BuildDelayFreeUses when register types match

a74nh commented:
Rebased on top of the other fixes. As mentioned in #107786, fixed it so that `BuildDelayFreeUses()` is only called for matching register types. Need to confirm that there are not spmi diffs

a74nh commented:
No asm diffs now:

```
❯ python3 ./src/coreclr/scripts/superpmi.py collect $CORE_ROOT/corerun "./artifacts/tests/coreclr/linux.arm64.Checked/JIT/HardwareIntrinsics/HardwareIntrinsics_Arm_ro/HardwareIntrinsics_Arm_ro.dll"
[08:27:14] ================ Logging to /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/superpmi.log
[08:27:14] SuperPMI collect
[08:27:14] SuperPMI JIT Path: /home/alahay01/dotnet/runtime_sve_api/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/libclrjit.so
[08:27:14] Collecting using command:
[08:27:14]   /home/alahay01/dotnet/runtime_sve_api/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/corerun ./artifacts/tests/coreclr/linux.arm64.Checked/JIT/HardwareIntrinsics/HardwareIntrinsics_Arm_ro/HardwareIntrinsics_Arm_ro.dll
[08:30:53] Merging MC files
[08:30:57] Copy base MCH file to final MCH file
[08:31:08] Creating TOC file
[08:31:10] Generated MCH file: /home/alahay01/dotnet/runtime_sve_api/linux.arm64.Checked.mch

❯ python3 ./src/coreclr/scripts/superpmi.py asmdiffs -mch_files /home/alahay01/dotnet/runtime_sve_api/linux.arm64.Checked.mch
[08:35:57] ================ Logging to /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/superpmi.1.log
[08:35:57] Using JIT/EE Version from jiteeversionguid.h: b75a5475-ff22-4078-9551-2024ce03d383
[08:35:58] Baseline hash: cdc8418a7f4e51b771db2ae7ee5cde5f479cde7e
[08:35:58] Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb/linux/arm64/Checked/libclrjit.so -> /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/basejit/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb.linux.arm64.Checked/libclrjit.so
Downloading 5.6/5.6 MB...
[08:35:59] Downloaded https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb/linux/arm64/Checked/libclrjit.so
[08:35:59] Using baseline /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/basejit/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb.linux.arm64.Checked/libclrjit.so
[08:35:59] Using coredistools found at /home/alahay01/dotnet/runtime_sve_api/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/libcoredistools.so
[08:35:59] SuperPMI ASM diffs
[08:35:59] Base JIT Path: /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/basejit/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb.linux.arm64.Checked/libclrjit.so
[08:35:59] Diff JIT Path: /home/alahay01/dotnet/runtime_sve_api/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/libclrjit.so
[08:35:59] Using MCH files:
[08:35:59]   /home/alahay01/dotnet/runtime_sve_api/linux.arm64.Checked.mch
[08:35:59] Running asm diffs of /home/alahay01/dotnet/runtime_sve_api/linux.arm64.Checked.mch
[08:36:39] Clean SuperPMI diff (72927 contexts processed)
[08:36:39] Asm diffs summary:
[08:36:39]   Summary Markdown file: /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/diff_summary.md
[08:36:39]   Short Summary Markdown file: /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/diff_short_summary.md
[08:36:39]   No asm diffs

❯ cat /home/alahay01/dotnet/runtime_sve_api/artifacts/spmi/diff_summary.md
Diffs are based on <span style="color:#1460aa">72,927</span> contexts (<span style="color:#1460aa">1</span> MinOpts, <span style="color:#1460aa">72,926</span> FullOpts).

No diffs found.

<details>
<summary>Details</summary>
<div style="margin-left:1em">

#### Context information

|Collection|Diffed contexts|MinOpts|FullOpts|Missed, base|Missed, diff|
|---|--:|--:|--:|--:|--:|
|linux.arm64.Checked.mch|72,927|1|72,926|0 (0.00%)|0 (0.00%)|




</div></details>
```

a74nh committed:
Merge main

a74nh committed:
Assert only on Arm64

kunalspathak reviewed (commented):
Added some minor comments.

a74nh committed:
Comment fixups

a74nh commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1385,818 +1385,428 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+
+        if (resultOpNum != 0)
+        {
+            embeddedDelayFreeOp = embeddedOpNode->Op(resultOpNum);
```
> hhm, can you confirm if we just overwrite the `embeddedDelayFreeOp` that was passed as the parameter to this function?

Yes. the value being passed into `BuildEmbeddedOperandUses()` is the address of `embeddedDelayFreeOp`. That value can be changed within `BuildEmbeddedOperandUses()` and it won't effect any variables outside `BuildEmbeddedOperandUses()` that were also same to the same value.
I think that's what you were asking for.

a74nh commented:
All fixed up as suggested.

kunalspathak commented:
I was hoping this to be zero diff, but there are still diffs related to `LoadAndInsertScalar` intrinsics. Can you please double check?

![image](https://github.com/user-attachments/assets/708e01e4-df5b-478b-970a-7989420bda5c)

a74nh commented:
> I was hoping this to be zero diff, but there are still diffs related to `LoadAndInsertScalar` intrinsics. Can you please double check?
> 

Looks like the reason for this is that in HEAD, `op1` is generated via:
```c++
            assert(isRMW);
            assert(intrin.op1->OperIs(GT_FIELD_LIST));

            GenTreeFieldList* op1 = intrin.op1->AsFieldList();
            assert(compiler->info.compNeedsConsecutiveRegisters);

            for (GenTreeFieldList::Use& use : op1->Uses())
            {
                BuildDelayFreeUses(use.GetNode(), intrinsicTree);
                srcCount++;
            }
```

My patch generates `op1` via:
```c++
            srcCount += BuildConsecutiveRegistersForUse(operand, delayFreeOp);
```

`BuildConsecutiveRegistersForUse()` was added for generating `op1` for `NI_AdvSimd_VectorTableLookup` and `NI_AdvSimd_Arm64_VectorTableLookup`.

AIUI,  `BuildConsecutiveRegistersForUse()` is a more generic function and so should be useable for `LoadAndInsertScalarVector...`

a74nh commented:
> > I was hoping this to be zero diff, but there are still diffs related to `LoadAndInsertScalar` intrinsics. Can you please double check?

With this patch there are no diffs again. It would be shame to put something so specific in this patch.

```
diff --git a/src/coreclr/jit/lsraarm64.cpp b/src/coreclr/jit/lsraarm64.cpp
index d124404beb0..6646791f790 100644
--- a/src/coreclr/jit/lsraarm64.cpp
+++ b/src/coreclr/jit/lsraarm64.cpp
@@ -1439,8 +1439,32 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
         {
             assert(candidates == RBM_NONE);

-            // Some operands have consective op which is also a delay free op
-            srcCount += BuildConsecutiveRegistersForUse(operand, delayFreeOp);
+            switch (intrin.id)
+            {
+                case NI_AdvSimd_LoadAndInsertScalarVector64x2:
+                case NI_AdvSimd_LoadAndInsertScalarVector64x3:
+                case NI_AdvSimd_LoadAndInsertScalarVector64x4:
+                case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x2:
+                case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x3:
+                case NI_AdvSimd_Arm64_LoadAndInsertScalarVector128x4:
+                {
+                    assert(consecutiveOp->OperIs(GT_FIELD_LIST));
+
+                    GenTreeFieldList* fieldOp = operand->AsFieldList();
+
+                    for (GenTreeFieldList::Use& use : fieldOp->Uses())
+                    {
+                        BuildDelayFreeUses(use.GetNode(), intrinsicTree);
+                        srcCount++;
+                    }
+                    break;
+                }
+
+                default:
+                    // Some operands have consective op which is also a delay free op
+                    srcCount += BuildConsecutiveRegistersForUse(operand, delayFreeOp);
+                    break;
+            }
         }
         else if (delayFreeOp == operand)
         {
```

kunalspathak commented:
> With this patch there are no diffs again. It would be shame to put something so specific in this patch.

Alright, I did some digging and turns out that for upperVectorRestore, we were not marking them as "delay-free", which this PR does.


![image](https://github.com/user-attachments/assets/aa61b0a3-463d-4f6e-9d8e-d1469012a829)

kunalspathak reviewed (approved):
LGTM
